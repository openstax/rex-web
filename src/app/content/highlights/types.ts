import { Highlight, HighlightColorEnum, HighlightsSummary, NewHighlight } from '@openstax/highlighter/dist/api';
import { ScrollTarget } from '../../navigation/types';
import { FiltersChange } from '../components/popUp/types';
import { LinkedArchiveTreeNode } from '../types';

export interface SummaryHighlights {
  [locationId: string]: {[pageId: string]: HighlightData[]};
}

export type OrderedSummaryHighlights = Array<{
  location: LinkedArchiveTreeNode,
  pages: Array<{
    pageId: string;
    highlights: HighlightData[];
  }>
}>;

export interface SummaryFilters {
  locationIds: string[];
  colors: HighlightColorEnum[];
}

export interface SummaryFiltersUpdate {
  locations?: FiltersChange<LinkedArchiveTreeNode>;
  colors?: FiltersChange<HighlightColorEnum>;
}

export type CountsPerSource = NonNullable<HighlightsSummary['countsPerSource']>;
export type HighlightColorCounts = CountsPerSource[string];

export type SummaryHighlightsPagination = null | {
  sourceIds: string[];
  page: number;
  perPage: number;
};

export interface State {
  currentPage: {
    pageId: string | null,
    highlights: null | HighlightData[];
    hasUnsavedHighlight: boolean;
    focused?: string;
  };
  summary: {
    open: boolean,
    pagination: SummaryHighlightsPagination,
    totalCountsPerPage: CountsPerSource | null;
    filters: SummaryFilters,
    loading: boolean;
    highlights: SummaryHighlights | null;
  };
}

export interface HighlightScrollTarget extends ScrollTarget {
  type: 'highlight';
  id: string;
}

// scopeId should not be nullable in the swagger
export type NewHighlightPayload = NewHighlight & {scopeId: string};
export type HighlightData = Highlight & {scopeId: string};
