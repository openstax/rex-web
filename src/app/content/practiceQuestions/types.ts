import { LinkedArchiveTreeSection } from '../types';

export interface State {
  summary: PracticeQuestionsSummary | null;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
  questionAnswers: QuestionAnswers;
  loading: boolean;
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

export interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export interface QuestionAnswers {
  [key: string]: PracticeAnswer | null;
}

export interface SetPracticeAnswer {
  questionId: string;
  answer: PracticeAnswer | null;
}
