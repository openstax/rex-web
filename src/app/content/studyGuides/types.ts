import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import {
  CountsPerSource,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from '../highlights/types';

export interface StudyGuidesSummaryFilters {
  colors: HighlightColorEnum[];
  locationIds: string[];
}

export interface State {
  isEnabled: boolean;
  summary: {
    filters: StudyGuidesSummaryFilters,
    loading: boolean,
    open: boolean,
    pagination: SummaryHighlightsPagination,
    studyGuides: SummaryHighlights | null
    totalCountsPerPage: CountsPerSource | null;
  };
}
