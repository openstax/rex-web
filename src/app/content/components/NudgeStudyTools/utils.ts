import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
