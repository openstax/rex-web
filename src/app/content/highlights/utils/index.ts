export {
  getHighlightLocationFilters,
  getHighlightColorFiltersWithContent,
  getHighlightLocationFilterForPage,
  getHighlightLocationFiltersWithContent,
  sectionIsHighlightLocationFilter
} from './locationFiltersUtils';

export {
  addToTotalCounts,
  updateInTotalCounts,
  removeFromTotalCounts,
  addSummaryHighlight,
  getHighlightByIdFromSummaryHighlights,
  removeSummaryHighlight,
  updateSummaryHighlight,
  updateSummaryHighlightsDependOnFilters,
  getSortedSummaryHighlights
} from './summaryHighlightsUtils';
