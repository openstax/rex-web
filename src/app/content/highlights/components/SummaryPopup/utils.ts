import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
