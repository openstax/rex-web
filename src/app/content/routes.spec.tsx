import createTestServices from '../../test/createTestServices';
import { book, page } from '../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../test/mocks/osWebLoader';
import { reactAndFriends, resetModules } from '../../test/utils';
import { Match } from '../navigation/types';

const longID = 'longidin-vali-dfor-mat1-111111111111';

describe('content route', () => {
  let content: any;
  let React: any; // tslint:disable-line:variable-name
  let renderer: any;
  let createApp: any;

  describe('generates urls', () => {
    it('for book slug and page slug', () => {
      content = require('./routes').content;
      const url = content.getUrl({book: {slug: 'book'}, page: {slug: 'page'}});
      expect(url).toEqual('/books/book/pages/page');
    });

    it('for book slug with version and page slug', () => {
      content = require('./routes').content;
      const url = content.getUrl({book: {slug: 'book', version: 'asdf'}, page: {slug: 'page'}});
      expect(url).toEqual('/books/book@asdf/pages/page');
    });

    it('for book uuid with version and page slug', () => {
      content = require('./routes').content;
      const url = content.getUrl({
        book: {
          uuid: longID,
          version: '1.0',
        },
        page: {
          slug: 'page',
        }});
      expect(url).toEqual('/books/longidin-vali-dfor-mat1-111111111111@1.0/pages/page');
    });

    it('for book slug and page uuid', () => {
      content = require('./routes').content;
      const url = content.getUrl({book: {slug: 'book'}, page: {uuid: longID}});
      expect(url).toEqual('/books/book/pages/longidin-vali-dfor-mat1-111111111111');
    });

    it('for book slug with version and page uuid', () => {
      content = require('./routes').content;
      const url = content.getUrl({book: {slug: 'book', version: 'asdf'}, page: {uuid: longID}});
      expect(url).toEqual('/books/book@asdf/pages/longidin-vali-dfor-mat1-111111111111');
    });

    it('for book uuid with version and page uuid', () => {
      content = require('./routes').content;
      const url = content.getUrl({
        book: {
          uuid: longID,
          version: '1.0',
        },
        page: {
          uuid: longID,
        }});
      expect(url).toEqual('/books/longidin-vali-dfor-mat1-111111111111@1.0/pages/longidin-vali-dfor-mat1-111111111111');
    });
  });

  describe('route renders', () => {
    const windowBackup = window;
    const documentBackup = document;

    beforeEach(() => {
      delete (global as any).window;
      delete (global as any).document;
      resetModules();
      ({React, renderer} = reactAndFriends());
      content = require('./routes').content;
      createApp = require('../index').default;
    });

    afterEach(() => {
      (global as any).window = windowBackup;
      (global as any).document = documentBackup;
    });

    it('renders a component', () => {
      const services = createTestServices();

      const params = {
        book: {
          slug: mockCmsBook.meta.slug,
        },
        page: {
          slug: 'test-page-1',
        },
      };
      const state = {
        bookUid: book.id,
        bookVersion: book.version,
        pageUid: page.id,
      };

      const match: Match<typeof content> = {
        params,
        route: content,
        state,
      };
      const app = createApp({
        initialEntries: [match],
        services,
      });

      const tree = renderer.create(<app.container />);

      expect(tree.toJSON()).toMatchSnapshot();

      tree.unmount();
    });
  });

  describe('route content', () => {
    const mockBookConfig = {
      [book.id]: {defaultVersion: book.version},
    } as {[key: string]: {defaultVersion: string}};

    it('doesnt set archive url on getSearch when there is only a default archive url', () => {
      resetModules();
      jest.doMock('../../config', () => ({REACT_APP_ARCHIVE_URL: 'some-content'}));
      jest.doMock('../../config.books', () => mockBookConfig);
      expect(require('./routes').content.getSearch()).toEqual('');
    });

    it('sets archive url on getSearch when there is an archive url override and no default', () => {
      resetModules();
      jest.doMock('../../config', () => ({REACT_APP_ARCHIVE_URL_OVERRIDE: 'some-content'}));
      jest.doMock('../../config.books', () => mockBookConfig);
      expect(require('./routes').content.getSearch()).toEqual('archive=some-content');
    });

    it('sets archive url on getSearch when there is an archive url override and a default', () => {
      resetModules();
      jest.doMock('../../config', () => ({
        REACT_APP_ARCHIVE_URL: 'asdf',
        REACT_APP_ARCHIVE_URL_OVERRIDE: 'some-content',
      }));
      jest.doMock('../../config.books', () => mockBookConfig);
      expect(require('./routes').content.getSearch()).toEqual('archive=some-content');
    });
  });
});
