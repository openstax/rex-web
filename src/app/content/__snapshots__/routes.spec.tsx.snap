// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assigned route route renders renders a component 1`] = `
Array [
  <a
    className="HiddenLink-wwwegq-0 bkbyvj"
    href="#main-content"
    onClick={[Function]}
  >
    Skip to Content
  </a>,
  <a
    className="HiddenLink-wwwegq-0 bkbyvj"
    href={
      Array [
        "https://openstax.org/accessibility-statement",
      ]
    }
  >
    Go to accessibility page
  </a>,
  <a
    className="HiddenLink-wwwegq-0 bkbyvj"
    href="#"
    onClick={[Function]}
  >
    Keyboard shortcuts menu
  </a>,
  <div
    className="styled__TopBarWrapper-sc-12dq39v-0 sc-bwzfXH efJRLb"
  >
    <h2
      className="sc-htpNat iadNqp"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<span class=\\"os-text\\">Test Page 1</span>",
        }
      }
    />
    <div
      className="styled__TextResizerDropdown-sc-12dq39v-16 iLqgWD Dropdown-rmc6yw-5 fBPFfC Dropdown__TabHiddenDropDown-rmc6yw-1 kbMrSv"
    >
      <button
        aria-label="Change text size"
        className="Dropdown__DropdownToggle-rmc6yw-0 gpiKKF Filters__Toggle-sc-1fn81fx-1 liapzj Button__PlainButton-ayg7nk-2 chgswm"
        data-analytics-label="Change text size"
        onClick={[Function]}
      >
        <div
          tabIndex={-1}
        >
          <img
            alt=""
            aria-hidden="true"
            src="text-size.svg"
          />
        </div>
      </button>
    </div>
  </div>,
  <div
    className="MinPageHeight-sc-1i28xs0-0 hxgCaZ"
  >
    <div
      className="RedoPadding-sc-1rj2rww-0 gaOlpC"
    >
      <div
        className="page-content PageContent-ny9bj0-0 bXhXti"
        tabIndex={0}
      >
        <div
          className="MainContent__ContentStyles-sc-6yy1if-0 evJIOJ"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<div xmlns:m=\\"http://www.w3.org/1998/Math/MathML\\" xmlns:ns2=\\"urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0\\" class=\\"introduction\\" data-cnxml-to-html-ver=\\"1.3.2\\" data-type=\\"page\\" id=\\"1d1fd537-77fb-4eac-8a8a-60bbaa747b6d\\"><p id=\\"p1\\">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi ex erat, lacinia vitae mattis id, commodo vitae est. Nam quis arcu quis enim congue aliquet. Vivamus dictum rhoncus tortor, malesuada placerat tortor imperdiet ac. Fusce ac viverra nisi. Aliquam ut tempor diam. Cras rhoncus sodales massa vitae porttitor. Integer pharetra lorem a ante sollicitudin condimentum. Aliquam imperdiet erat et tellus mattis finibus. Sed nisl lectus, ultricies et ante ut, porttitor varius sem. Interdum et malesuada fames ac ante ipsum primis in faucibus. Fusce tempus interdum ante, nec placerat felis facilisis quis. Nunc nec malesuada nisi. Cras eu iaculis felis. Sed id maximus enim, sit amet sodales sapien. Proin pulvinar sem risus, nec ultricies odio viverra a. Aliquam feugiat euismod dapibus.</p><section id=\\"section1\\"><h3 data-type=\\"document-title\\" id=\\"100_copy_1\\"><span class=\\"os-number\\">1.1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">A Section, Probably</span></h3><p id=\\"p2\\">Morbi lobortis mattis velit, <math display=\\"inline\\"><semantics><mrow><mrow><mrow><mrow><mrow><mrow><mi>W</mi><mo stretchy=\\"false\\">=</mo><mrow><msub><mtext>KE</mtext><mrow><mn>f</mn></mrow></msub><mo stretchy=\\"false\\">+</mo><msub><mtext>PE</mtext><mrow><mn>g</mn></mrow></msub></mrow></mrow><mo stretchy=\\"false\\">=</mo><mfrac><mrow><mn>1</mn></mrow><mrow><mn>2</mn></mrow></mfrac><msup><msub><mi fontstyle=\\"italic\\">mv</mi><mn>f</mn></msub><mn>2</mn></msup><mo stretchy=\\"false\\">+</mo><mrow><mtext fontstyle=\\"italic\\">mgh</mtext></mrow></mrow></mrow></mrow><mrow></mrow></mrow></mrow><annotation-xml encoding=\\"MathML-Content\\"><semantics><mrow><mrow><mrow><mrow><mrow><mi>W</mi><mo stretchy=\\"false\\">=</mo><mrow><msub><mtext>KE</mtext><mrow><mn>f</mn></mrow></msub><mo stretchy=\\"false\\">+</mo><msub><mtext>PE</mtext><mrow><mn>g</mn></mrow></msub></mrow></mrow><mo stretchy=\\"false\\">=</mo><mfrac><mrow><mn>1</mn></mrow><mrow><mn>2</mn></mrow></mfrac><msup><msub><mi fontstyle=\\"italic\\">mv</mi><mn>f</mn></msub><mn>2</mn></msup><mo stretchy=\\"false\\">+</mo><mrow><mtext fontstyle=\\"italic\\">mgh</mtext></mrow></mrow></mrow></mrow><mrow></mrow></mrow><annotation encoding=\\"StarMath 5.0\\"> size 12{W=\\"KE\\" rSub { size 8{f} } +\\"PE\\" rSub { size 8{g} } = {  { size 8{1} }  over  { size 8{2} } }  ital \\"mv\\" rSub { size 8{f}   rSup { size 8{2} } } + ital \\"mgh\\"} {}</annotation></semantics></annotation-xml></semantics></math> dapibus convallis est mollis sed. Sed ullamcorper est tortor, at ultricies felis tincidunt ut. Mauris interdum nunc et convallis pharetra. Nulla ac erat nulla. Vivamus a ornare leo. Quisque luctus dui eget auctor tristique. Sed vel est eget lorem volutpat mattis non et nisi. Aliquam ultricies ultrices velit, id fringilla ex suscipit in. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris consequat blandit felis, non finibus neque dignissim sit amet. Integer laoreet dapibus quam eget pulvinar. Aenean lobortis, arcu dignissim euismod ornare, eros est tincidunt urna, nec auctor odio quam eget tortor. Praesent sit amet metus a metus varius tincidunt. Pellentesque convallis sit amet erat eget malesuada. Proin pretium est euismod risus facilisis dignissim. Interdum et malesuada fames ac ante ipsum primis in faucibus.</p><p id=\\"p3\\">Integer maximus augue vitae orci vehicula tincidunt. Integer auctor ante odio, nec porta nisi pellentesque id. Donec porttitor velit vel turpis sagittis molestie. Sed eu rhoncus orci. Donec feugiat tristique posuere. Fusce consequat, nisi vitae porttitor ornare, augue velit sodales augue, ac facilisis turpis orci vitae lectus. Mauris mauris urna, dictum a dui et, imperdiet commodo justo. Pellentesque scelerisque dui sed libero porta, sit amet rhoncus ex dictum. Interdum et malesuada fames ac ante ipsum primis in faucibus.</p><div data-type=\\"equation\\" id=\\"fs-id2746082\\"><math display=\\"block\\"><semantics><mrow><mrow><mrow><mrow><mi>P</mi><mo stretchy=\\"false\\">=</mo><mfrac><mi>W</mi><mi>t</mi></mfrac></mrow></mrow><mrow></mrow></mrow></mrow><annotation-xml encoding=\\"MathML-Content\\"><semantics><mrow><mrow><mrow><mi>P</mi><mo stretchy=\\"false\\">=</mo><mfrac><mi>W</mi><mi>t</mi></mfrac></mrow></mrow><mrow></mrow></mrow><annotation encoding=\\"StarMath 5.0\\"> size 12{P= {  {W}  over  {t} } } {}</annotation></semantics></annotation-xml></semantics></math><div class=\\"os-equation-number\\"><span class=\\"os-number\\">1.01</span></div></div><p id=\\"p4\\">Donec congue tortor sed magna ullamcorper mollis. Suspendisse eu tincidunt augue, et pulvinar purus. Fusce gravida quam lorem, sed pharetra libero bibendum et. Etiam facilisis ex vel nisi egestas, non consequat mi malesuada. Donec scelerisque magna sem, ut ornare massa rhoncus sed. In ultricies tellus porttitor enim elementum, ut mollis nisl euismod. Pellentesque eu lobortis neque. Phasellus eget luctus velit, in eleifend lectus. Pellentesque leo odio, cursus eget feugiat eu, finibus in nibh. In nec urna justo. Donec ut nisi eleifend, auctor lacus quis, hendrerit sem. Pellentesque sagittis nunc a molestie finibus. Donec pulvinar nibh elit, nec ornare nibh egestas quis. Curabitur luctus euismod lobortis.</p><div class=\\"os-figure\\" id=\\"figureid\\"><figure data-id=\\"figureid\\"><div data-alt=\\"\\" data-type=\\"media\\" id=\\"Phet_module_25.2\\"> <iframe title=\\"interactive simulation\\" height=\\"371.4\\" width=\\"660\\" data-lazy-src=\\"/specials/version/radio-waves/#sim-radio-waves\\"></iframe></div></figure><div class=\\"os-caption-container\\"><span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.0</span><span class=\\"os-divider\\"> </span></div></div>  <p id=\\"p5\\">Maecenas vehicula nec quam vitae sodales. Vestibulum bibendum accumsan mauris. Nunc rhoncus libero odio, in ullamcorper massa ornare convallis. Nullam metus tortor, aliquam a ultrices et, rhoncus in risus. Nunc vel turpis erat. Cras semper sagittis odio, eu mattis nulla faucibus quis. Aliquam egestas nunc elit, eget condimentum lectus congue et. Curabitur non gravida lorem. Integer mi nunc, commodo vitae augue vel, bibendum pretium magna. Sed ut nibh venenatis ipsum tempus mollis pretium eget est. Aliquam commodo quam ipsum, at fermentum enim tristique eget. Mauris imperdiet dapibus maximus. Donec tristique varius lacinia. Pellentesque pharetra nunc eu nisi ullamcorper placerat. Suspendisse potenti.</p><p id=\\"p6\\">Integer fringilla blandit dolor ut mattis. Nam commodo ex at blandit sodales. Nullam suscipit, nunc eu mollis sagittis, purus arcu tincidunt ligula, vitae pharetra metus tortor id ex. Donec malesuada aliquet ligula id mollis. Maecenas scelerisque maximus sollicitudin. Quisque sagittis dolor at ligula congue, quis rhoncus mi porttitor. Quisque lobortis ut ante sed convallis. Integer luctus cursus molestie.</p><p id=\\"p7\\">Morbi a purus a elit dapibus fermentum. Sed varius ex quis tortor tincidunt posuere. Mauris luctus risus id aliquet mollis. Phasellus ornare leo a arcu imperdiet dictum. Praesent accumsan venenatis urna, sed egestas sem pharetra a. Sed vulputate sit amet nisl a aliquet. Interdum et malesuada fames ac ante ipsum primis in faucibus. Mauris facilisis augue vitae est maximus pharetra in ac elit. Pellentesque viverra purus non ligula tempor convallis.</p><p id=\\"p8\\">Pellentesque lectus ante, malesuada ut ornare vitae, euismod non urna. Morbi varius eros eget felis accumsan vehicula. In diam ex, finibus rhoncus dapibus non, lacinia in lacus. Nullam pretium mollis laoreet. Ut interdum dui sit amet elementum dignissim. Etiam in nibh vitae ligula fringilla aliquet quis vel ex. Nullam quis elit libero. In aliquet laoreet arcu at posuere. Duis consectetur et dolor non accumsan. Fusce enim est, dignissim sed mauris at, ornare cursus ligula. Nullam tincidunt eu purus vel lacinia. Praesent varius, arcu sed facilisis elementum, orci metus vestibulum nisl, sed sodales urna eros ullamcorper turpis. Donec mollis ac massa ac elementum. Nulla maximus purus sed orci efficitur, ut porta mauris elementum.</p><p id=\\"p9\\">Vestibulum blandit enim quis sapien ultrices pretium. Cras dapibus ornare fringilla. Duis ut laoreet dolor. Proin ullamcorper molestie dui eu vestibulum. Nullam convallis mollis mauris nec accumsan. Proin egestas enim non nisl suscipit, ut feugiat felis pharetra. In faucibus fringilla ipsum vel sollicitudin. Maecenas blandit augue odio, ac mattis justo feugiat sit amet. Etiam nec molestie eros. Sed gravida velit libero, at pretium diam venenatis sit amet. Donec tellus leo, faucibus eu sapien nec, mollis efficitur lorem. Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p></section><dl id=\\"test-pair-page1\\"><dt>test term</dt><dd>test definition with <math>math element</math></dd></dl></div>",
            }
          }
          data-dynamic-style={false}
          id="main-content"
          tabIndex={-1}
          textSize={0}
        />
      </div>
    </div>
  </div>,
  <details
    className="Details-xi326o-3 Attribution__AttributionDetails-sc-11isnv6-4 ecFdAU"
    data-analytics-region="attribution"
    data-testid="attribution-details"
  >
    <summary
      aria-label="Citation/Attribution"
      className="Attribution__AttributionSummary-sc-11isnv6-2 ccMXsL Details__Summary-xi326o-2 goqLfs"
    >
      <svg
        aria-hidden="true"
        className="Attribution__SummaryClosedIcon-sc-11isnv6-0 gSPJtD Details__ExpandIcon-xi326o-0 cVuUUi StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 192 512"
      >
        <path
          d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        className="Attribution__SummaryOpenIcon-sc-11isnv6-1 kRXAve Details__CollapseIcon-xi326o-1 jsdrhW StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 320 512"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
      <span>
        Citation/Attribution
      </span>
    </summary>
    <div
      className="Attribution__Content-sc-11isnv6-3 SNfN"
      dangerouslySetInnerHTML={
        Object {
          "__html": "
<p>
  Want to cite, share, or modify this book? This book uses the
  <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"http://example.com\\">
    Sweet License 1.0
  </a> and you must attribute OpenStax.
</p>

<strong>Attribution information</strong>
<ul>
  <li>
    If you are redistributing all or part of this book in a print format,
    then you must include on every physical page the following attribution:
    <blockquote>
      Access for free at https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units
    </blockquote>
  </li>
  <li>
    If you are redistributing all or part of this book in a digital format,
    then you must include on every digital page view the following attribution:
    <blockquote>
      Access for free at <a href=\\"https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
    </blockquote>
  </li>
</ul>

<strong>Citation information</strong>
<ul>
  <li>
    Use the information below to generate a citation. We recommend using a
    citation tool such as
    <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.lib.ncsu.edu/citationbuilder/#/default/default\\">this one</a>.
    <ul>
      <li>
        Authors: Bam Bammerson
      </li>
      <li>
        Publisher/website: OpenStax
      </li>
      <li>
        Book title: Test Book 1
      </li>
      <li>
        Publication date: Jun 21, 2012
      </li>
      <li>
        Location: Houston, Texas
      </li>
      <li>
        Book URL: <a href=\\"https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
      </li>
      <li>
        Section URL: <a href=\\"https://openstax.org/books/book-slug-1/pages/test-page-1\\">https://openstax.org/books/book-slug-1/pages/test-page-1</a>
      </li>
    </ul>
  </li>
</ul>

<p>
  © Jun 21, 2012 OpenStax. Textbook content produced by OpenStax is licensed under a Sweet License 1.0. <strong>The OpenStax name, OpenStax logo, OpenStax book covers, OpenStax CNX name, and OpenStax CNX logo
  are not subject to the Creative Commons license and may not be reproduced without the prior and express written
  consent of Rice University.</strong>
</p>
",
        }
      }
    />
  </details>,
]
`;

exports[`assigned route route renders renders loading state (for loadable component) 1`] = `null`;

exports[`content route route renders renders a component 1`] = `null`;
